In [1]:
#import library
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler


In [8]:
#loading the dataset
df=pd.read_csv('/kaggle/input/data-for-investing-type-prediction/investing_program_prediction_data.csv')
#showing the dataset
df

SE1  SE2  BA1  BA2    BA3      BA4      BA5      BA6     BA7 PE1 PE2  \
0      45  G29   12  0.0   5934        0        0        0       0  I0  I0   
1      41   G0    0  0.0  21514   316374     5142        0    5141  I0  I0   
2      43   G0    1  0.0  14324    22344    22344        0   22344  I0  I0   
3      36  G52   16  0.0  41114  2406838  1156219   692875   41425  I0  I0   
4       2  G10    0  0.0   2394      621      621        0     621  I0  I0   
...   ...  ...  ...  ...    ...      ...      ...      ...     ...  ..  ..   
4729   57  G23    1  0.0  12514    14274    14274        0   14274  I0  I0   
4730   34  G66    0  0.0   5994     1036     1036        0    1036  I0  I0   
4731   44  G51   19  0.0  69627  4805913  3345950  1899126  596537  I0  I0   
4732   22   G0    0  0.0   5999        0        0        0       0  I0  I0   
4733   47  G67    0  0.0  23531    16786    16786        0   16786  I0  I0   

     PE3 PE4 PE5 PE6 PE7 PE8 PE9 PE10 PE11 PE12 PE13 PE14 PE15  IA1  IA2  IA3  \
0     I0  I0  I0  I0  I0  I0  I0   I0   I0   I0   I0   I0   I0    0    0    1   
1     I0  I0  I0  I0  I0  I0  I0   I0   I0   I1   I1   I1   I0    0    0    1   
2     I0  I0  I0  I0  I0  I0  I0   I0   I0   I0   I0   I0   I0    0    0    0   
3     I0  I0  I1  I1  I1  I0  I0   I1   I1   I1   I1   I0   I0    0    0    0   
4     I0  I0  I0  I0  I0  I0  I0   I0   I0   I0   I0   I0   I0    0    0    0   
...   ..  ..  ..  ..  ..  ..  ..  ...  ...  ...  ...  ...  ...  ...  ...  ...   
4729  I0  I0  I0  I0  I0  I0  I0   I0   I0   I0   I0   I0   I0    0    0    0   
4730  I0  I0  I0  I0  I0  I0  I0   I0   I0   I0   I0   I0   I0    0    0    0   
4731  I0  I1  I1  I0  I1  I0  I0   I1   I1   I1   I1   I1   I0    0    0    1   
4732  I0  I0  I0  I0  I0  I0  I0   I0   I0   I0   I0   I0   I0    0    0    0   
4733  I0  I0  I0  I0  I0  I0  I0   I0   I0   I0   I0   I0   I0    0    0    0   

     InvType  
0         C1  
1         C1  
2         C1  
3         C1  
4         C1  
...      ...  
4729      C0  
4730      C1  
4731      C1  
4732      C1  
4733      C0  

[4734 rows x 28 columns]

# Preprocessing Function

In [18]:
def onehot_encode(df,column):
    df=df.copy()

    dummies=pd.get_dummies(df[column],prefix=column)
    df=pd.concat([df,dummies],axis=1)
    df=df.drop(column,axis=1)
    for column in df.columns:
        if df[column].dtypes=='bool':
            df[column]=df[column].astype(int)
    return df
    

In [36]:
def preprocess_inputs(df):
    df=df.copy()

    #removing non-numeric features 

    for column in ['PE'+str(i) for i in range(1,16)]:
        df[column]=df[column].apply(lambda x:x[1]).astype(int)

    #feature engineering the SE1 column
    df['Child']=df['SE1'].apply(lambda x:1 if x<18 else 0)
    df['Senior']=df['SE1'].apply(lambda x:1 if x>65 else 0)
    
    #onehot encoding SE2 column

    df=onehot_encode(df,'SE2')
    
    #Splitting between target and features column

    y=df['InvType']
    x=df.drop('InvType',axis=1)


    x_train,x_test,y_train,y_test=train_test_split(x,y,train_size=0.7,random_state=123,shuffle=True)
    scaler=StandardScaler()
    scaler.fit(x_train)
    x_train=pd.DataFrame(scaler.transform(x_train),columns=x_train.columns,index=x_train.index)
    x_test=pd.DataFrame(scaler.transform(x_test),columns=x_test.columns,index=x_test.index)












    
    
    return x_train,x_test,y_train,y_test

In [37]:
x_train,x_test,y_train,y_test=preprocess_inputs(df)
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(3313, 118)
(1421, 118)
(3313,)
(1421,)


In [35]:
x_train

SE1       BA1       BA2       BA3       BA4       BA5       BA6  \
1448  0.284841 -0.667328 -0.075189 -0.379713 -0.486831 -0.491989 -0.288410   
2311 -0.422227 -0.667328 -0.075189 -0.374486 -0.483101 -0.484134 -0.288410   
1413  1.634698  3.009943 -0.075189 -0.310761 -0.469085 -0.454621 -0.288410   
4044  1.056188  0.149843 -0.075189  0.091508  0.426379  1.218070 -0.288410   
3158 -1.322131 -0.667328 -0.075189 -0.379713 -0.487118 -0.492594 -0.288410   
...        ...       ...       ...       ...       ...       ...       ...   
1593 -0.936458 -0.122547 -0.075189 -0.345823 -0.482573 -0.483023 -0.288410   
4060  0.734793 -0.531133 -0.075189 -0.073772 -0.419290 -0.349762  0.004929   
1346 -0.036553 -0.531133 -0.075189 -0.368497 -0.480164 -0.477949 -0.288410   
3454 -1.322131 -0.667328 -0.075189 -0.086108 -0.418123 -0.347304 -0.288410   
3582 -0.486506 -0.667328 -0.075189  0.355138 -0.308952 -0.117410 -0.288410   

           BA7       PE1       PE2       PE3       PE4       PE5       PE6  \
1448 -0.425472 -0.231124 -0.055023 -0.090646 -0.085423 -0.664124 -0.383615   
2311 -0.406809 -0.231124 -0.055023 -0.090646 -0.085423 -0.664124 -0.383615   
1413 -0.336685 -0.231124 -0.055023 -0.090646 -0.085423 -0.664124 -0.383615   
4044  0.031984 -0.231124 -0.055023 -0.090646 -0.085423  1.505742  2.606778   
3158 -0.426910 -0.231124 -0.055023 -0.090646 -0.085423 -0.664124 -0.383615   
...        ...       ...       ...       ...       ...       ...       ...   
1593 -0.404168 -0.231124 -0.055023 -0.090646 -0.085423  1.505742 -0.383615   
4060 -0.087534 -0.231124 -0.055023 -0.090646 -0.085423 -0.664124 -0.383615   
1346 -0.392113 -0.231124 -0.055023 -0.090646 -0.085423 -0.664124 -0.383615   
3454 -0.081694 -0.231124 -0.055023 -0.090646 -0.085423 -0.664124 -0.383615   
3582  0.464544 -0.231124 -0.055023 -0.090646 -0.085423 -0.664124 -0.383615   

           PE7       PE8      PE9      PE10     PE11      PE12      PE13  \
1448 -0.271857 -0.289995 -0.21394 -0.658467 -0.51383 -0.502074 -0.602806   
2311 -0.271857 -0.289995 -0.21394 -0.658467 -0.51383 -0.502074 -0.602806   
1413 -0.271857 -0.289995 -0.21394 -0.658467 -0.51383 -0.502074 -0.602806   
4044 -0.271857 -0.289995 -0.21394  1.518679 -0.51383  1.991737  1.658910   
3158 -0.271857 -0.289995 -0.21394 -0.658467 -0.51383 -0.502074 -0.602806   
...        ...       ...      ...       ...      ...       ...       ...   
1593 -0.271857 -0.289995 -0.21394 -0.658467 -0.51383 -0.502074 -0.602806   
4060 -0.271857 -0.289995 -0.21394 -0.658467 -0.51383 -0.502074 -0.602806   
1346 -0.271857 -0.289995 -0.21394 -0.658467 -0.51383 -0.502074 -0.602806   
3454 -0.271857 -0.289995 -0.21394 -0.658467 -0.51383 -0.502074 -0.602806   
3582 -0.271857 -0.289995 -0.21394 -0.658467 -0.51383 -0.502074 -0.602806   

          PE14      PE15      IA1       IA2       IA3     Child    SE2_G0  \
1448 -0.294860 -0.249398 -0.03034 -0.017376 -0.522061 -0.238282  1.699558   
2311 -0.294860 -0.249398 -0.03034 -0.017376 -0.522061 -0.238282 -0.588388   
1413 -0.294860 -0.249398 -0.03034 -0.017376 -0.522061 -0.238282 -0.588388   
4044  3.391443 -0.249398 -0.03034 -0.017376  1.938234 -0.238282  1.699558   
3158 -0.294860 -0.249398 -0.03034 -0.017376 -0.522061 -0.238282  1.699558   
...        ...       ...      ...       ...       ...       ...       ...   
1593 -0.294860 -0.249398 -0.03034 -0.017376 -0.522061 -0.238282  1.699558   
4060 -0.294860 -0.249398 -0.03034 -0.017376 -0.522061 -0.238282 -0.588388   
1346 -0.294860 -0.249398 -0.03034 -0.017376 -0.522061 -0.238282 -0.588388   
3454 -0.294860 -0.249398 -0.03034 -0.017376 -0.522061 -0.238282 -0.588388   
3582 -0.294860 -0.249398 -0.03034 -0.017376 -0.522061 -0.238282  1.699558   

       SE2_G10   SE2_G11   SE2_G12   SE2_G13   SE2_G14   SE2_G15   SE2_G16  \
1448 -0.087198 -0.042595 -0.077933 -0.062765 -0.060293 -0.062765 -0.071818   
2311 -0.087198 -0.042595 -0.077933 -0.062765 -0.060293 -0.062765 -0.071818   
1413 -0.087198 -0.042595 -0.077933 -0.062765 -0.060293 -0.062765 

In [29]:
y_train.unique()

array(['C1', 'C0'], dtype=object)

# Training the Model

In [38]:
model=LogisticRegression()
model.fit(x_train,y_train)
print('The accuracy of Logistic Model',model.score(x_test,y_test))


The accuracy of Logistic Model 0.6882477128782547


# Without feature engineering accuracy 0.6678395496129487# 
# With feature engineering accuracy  0.6882477128782547

In [11]:
#getting preliminary information about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4734 entries, 0 to 4733
Data columns (total 28 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   SE1      4734 non-null   int64  
 1   SE2      4734 non-null   object 
 2   BA1      4734 non-null   int64  
 3   BA2      4734 non-null   float64
 4   BA3      4734 non-null   int64  
 5   BA4      4734 non-null   int64  
 6   BA5      4734 non-null   int64  
 7   BA6      4734 non-null   int64  
 8   BA7      4734 non-null   int64  
 9   PE1      4734 non-null   object 
 10  PE2      4734 non-null   object 
 11  PE3      4734 non-null   object 
 12  PE4      4734 non-null   object 
 13  PE5      4734 non-null   object 
 14  PE6      4734 non-null   object 
 15  PE7      4734 non-null   object 
 16  PE8      4734 non-null   object 
 17  PE9      4734 non-null   object 
 18  PE10     4734 non-null   object 
 19  PE11     4734 non-null   object 
 20  PE12     4734 non-null   object 
 21  PE13     4734 

In [13]:
#checking for missing value in the dataset
df.isna().sum().sum()

0